### PR TITLE
perf(moe_ep): CuTe-DSL 4.5.2 mainloop WAR — drop the 4.6.1 runtime floor

### PR DESCRIPTION
**Pull request by mhoqueanik**
## Summary

`nvidia-cutlass-dsl==4.5.2` codegen has a regression that makes the cutedsl MegaMoE nvfp4 swap-AB fc12 kernel 34–54% slower than 4.6.1 at every token count (documented in `kernel_src/cutedsl_megamoe/TUNING.md`, "CuTe-DSL runtime sensitivity"). This mattered because vLLM 0.25.1 pins exactly 4.5.2, forcing integrations to carry a 4.6.1 force-upgrade plus its compat chain (quack, tvm-ffi, tilelang, vendored-kernel patches).

This PR ports cutedsl_megamoe MR!27 (single file: `src/moe_nvfp4_swapab/kernel_fc12.py`): when the installed DSL is exactly 4.5.2, the MMA-consumer k-tile mainloop is peeled by one iteration (unconditional `try_wait` inside the loop, last tile issued after it). The gate is `cutlass.const_expr`-folded at trace time, so on any other version the generated kernel is byte-identical to before. 

## Version support matrix (all measured, 4x GB200, default geometry)

| DSL version | nvfp4 µs @1024/2048/8192 | status |
|---|---|---|
| 4.5.0 | — | fails at `cute.compile` (unsupported) |
| 4.5.1 | — | unmeasured; shim warns (<4.5.2) |
| 4.5.2 without WAR | 583.7 / 878.6 / 2579.5 | the regression |
| **4.5.2 + this WAR** | **428.6 / 621.5 / 1896.4** | full parity |
| 4.5.3 | 424.4 / 613.3 / 1933.3 | natively fast (upstream fix) |
| 4.6.0 | 427.0 / 614.1 / 1922.0 | natively fast |
| 4.6.1 | 428.5 / 625.6 / 1923.5 | reference |

Support statement: **>= 4.5.2 at full performance; < 4.5.2 unsupported.** The regression existed only in 4.5.2 and was fixed upstream in 4.5.3, so the exact `== 4.5.2` gate never affects any other version.

## Validation (all on pinned 4.5.2, 2026-07-22)

- **Correctness:** full `tests/moe_ep/run_tests.sh all` — all 8 sections   PASS (unit, torch-oracle, split multirank, bf16/nvfp4/ht correctness, mega multirank, smoke). 
- **Microbenchmark:** full 5-variant × 7-token-point reference sweep reproduces the 4.6.1 reference within run noise at every cell; adopted as the new TUNING.md reference table. mxfp8 needs no WAR (never regressed — measured).
- **vLLM 0.25.1 e2e (DeepSeek-V4-Flash, 4x GB200):** headline pair reproduces on vLLM's own 4.5.2 pin with no force-upgrade: prefill-8k **1.176x** native (53,623 vs 45,582 tok/s), decode-1k **1.068x** (34,263 vs 32,086 total tok/s) — within 0.7% of the 4.6.1-stack references.

## Changes

- `kernel_src/cutedsl_megamoe/src/moe_nvfp4_swapab/kernel_fc12.py`: the MR!27 loop peel, `const_expr`-gated on `== 4.5.2`.
- `kernel_src/cutedsl_megamoe/shim/__init__.py`: perf-floor warning now fires only below 4.5.2, with the measured support matrix in the docstring.
- `kernel_src/cutedsl_megamoe/TUNING.md`: 2026-07-15 sensitivity section marked OBSOLETE (kept as record); reference tables re-measured on 4.5.2 and adopted; version-gap measurements and support statement added.

AI-assisted (Claude Code): MR port, benchmark reruns, and docs.

## Summary by CodeRabbit

- **Performance**
  - Improved NVFP4 MoE behavior for supported CuTeDSL versions, including corrected 4.5.2 handling.
  - Refreshed microbenchmark and vLLM end-to-end results, with clarified throughput metrics and updated guidance on valid “performance floor” claims.

- **Compatibility**
  - Added an import-time warning for outdated CuTeDSL versions, with an environment variable to suppress it.

- **Bug Fixes**
  - Fixed zero-token routing/staging to correctly reset masked outputs in both fused and non-fused paths.

- **Documentation / Tests**
  - Expanded the benchmarking runbook and tuning instructions; added a GPU regression test for zero-token masking.

### CODE REVIEW

**Review by Anerudhan — APPROVED** [review]

File: flashinfer/moe_ep/kernel_src/cutedsl_megamoe/src/moe_nvfp4_swapab/moe_utils.py [hunk 1]
```diff
@@ -45,8 +45,9 @@
 
 from cutlass.utils.blockscaled_layout import tile_atom_to_shape_SF
 
-TensormapDescBytes = 128
-TensormapDescBytes = 64  # {$nv-internal-release}
+# Per-descriptor slot stride and alignment of the TMA descriptors held in the
+# tensormap workspace (see TensormapWorkspace below).
+TensormapDescBytes = 64
```
[the remark this answers is not in the document]
Comment (mhoqueanik) [reply flashinfer/moe_ep/kernel_src/cutedsl_megamoe/src/moe_nvfp4_swapab/moe_utils.py:58]: Addressed in last commit

File: tests/moe_ep/test_fused_quant_stage.py [hunk 2]
```diff
@@ -211,6 +211,70 @@ def _dg_buffers():
...
+    ``staged_tokens()`` reporting the stale count.
+    """
+    import torch
+
+    from flashinfer.moe_ep.kernel_src.cutedsl_megamoe import (
+        fused_quant_stage,
+        staged_tokens,
+    )
+
+    _require_blackwell()
```
[the remark this answers is not in the document]
Comment (mhoqueanik) [reply tests/moe_ep/test_fused_quant_stage.py:231]: cap[0] != 10 already skips SM120 (major 12), and SM103 is intentionally admitted — the fused staging kernel supports sm_100 and sm_103, as the skip message states. is_cvt_rs_supported guards an unrelated PTX instruction (mamba cvt.rs) and only coincidentally matches this arch set, so we don't want to couple to it. The CUDA-version semantics in is_sm100a_supported apply to nvcc targeting sm_100a; this test runs a CuTe-DSL kernel whose toolchain availability is already gated by the importorskip above. Resolving as no change needed.

**Review by aleozlx — APPROVED** [review]
